File: launcher/tests/main/update-resume.test.ts
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { afterEach, describe, expect, test, vi } from "vitest";
import {
  completeUpdateHeartbeat,
  consumeLauncherEntryProcessId,
  consumeUpdateHeartbeatEnvironment,
  launchInterruptedUpdateRecovery,
} from "@main/services/update-resume";
import { createLauncherSnapshot } from "../helpers/snapshot";

const roots: string[] = [];

async function tempRoot(label: string) {
  const root = await fs.mkdtemp(path.join(os.tmpdir(), `raylea-resume-${label}-`));
  roots.push(root);
  return root;
}

afterEach(async () => {
  await Promise.all(roots.splice(0).map((root) => fs.rm(root, { recursive: true, force: true })));
});

describe("update resume", () => {
  test("accepts and consumes the supervised launcher entry process id", () => {
    const environment = { RAYLEA_LAUNCHER_ENTRY_PID: "314" };

    expect(consumeLauncherEntryProcessId(environment, 271, 314)).toBe(314);
    expect(environment).not.toHaveProperty("RAYLEA_LAUNCHER_ENTRY_PID");
  });

  test("ignores launcher entry process ids that do not match the parent process", () => {
    const environment = { RAYLEA_LAUNCHER_ENTRY_PID: "999" };

    expect(consumeLauncherEntryProcessId(environment, 271, 314)).toBe(271);
    expect(environment).not.toHaveProperty("RAYLEA_LAUNCHER_ENTRY_PID");
  });

  test("consumes heartbeat secrets from the process environment", () => {
    const environment = {
      RAYLEA_UPDATE_HEARTBEAT: "C:/update/heartbeat.json",
      RAYLEA_UPDATE_TOKEN: "a".repeat(64),
      RAYLEA_UPDATE_RESTART_SERVICE: "true",
    };
    const request = consumeUpdateHeartbeatEnvironment(environment);

    expect(request).toEqual({
      heartbeatPath: "C:/update/heartbeat.json",
      token: "a".repeat(64),
      restartService: true,
    });
    expect(environment).not.toHaveProperty("RAYLEA_UPDATE_TOKEN");
  });

  test("writes a ready heartbeat only after service health and readiness pass", async () => {
    const parent = await tempRoot("heartbeat");
    const installRoot = path.join(parent, "RayleaBot");
    const transactionRoot = path.join(parent, ".rayleabot-update-test");
    await fs.mkdir(installRoot);
    await fs.mkdir(transactionRoot);
    await fs.writeFile(path.join(installRoot, "build_info.json"), JSON.stringify({
      version: "1.1.0",
      artifact_id: "windows-x64-full",
      update_protocol_version: 2,
    }));
    const heartbeatPath = path.join(transactionRoot, "postflight-launcher-heartbeat.json");
    const startService = vi.fn(async () => undefined);
    await completeUpdateHeartbeat(installRoot, {
      heartbeatPath,
      token: "b".repeat(64),
      restartService: true,
    }, {
      startService,
      getSnapshot: () => createLauncherSnapshot({
        launcher: { processLifecycle: "running" },
        server: { health: { status: "ok" }, readiness: { status: "ready", checks: {} } },
      }),
      launcherPid: 42,
      servicePid: () => 84,
    });

    const heartbeat = JSON.parse(await fs.readFile(heartbeatPath, "utf8")) as Record<string, unknown>;
    expect(startService).toHaveBeenCalledTimes(1);
    expect(heartbeat).toMatchObject({
      status: "ready",
      version: "1.1.0",
      launcher_pid: 42,
      service_pid: 84,
      service_running: true,
    });
  });

  test("reports failed postflight when readyz does not pass", async () => {
    const parent = await tempRoot("failed");
    const installRoot = path.join(parent, "RayleaBot");
    const transactionRoot = path.join(parent, ".rayleabot-update-test");
    await fs.mkdir(installRoot);
    await fs.mkdir(transactionRoot);
    await fs.writeFile(path.join(installRoot, "build_info.json"), JSON.stringify({
      version: "1.1.0",
      artifact_id: "windows-x64-full",
      update_protocol_version: 2,
    }));
    const heartbeatPath = path.join(transactionRoot, "postflight-launcher-heartbeat.json");
    await completeUpdateHeartbeat(installRoot, { heartbeatPath, token: "c".repeat(64), restartService: true }, {
      startService: async () => undefined,
      getSnapshot: () => createLauncherSnapshot({
        launcher: { processLifecycle: "running" },
        server: { health: { status: "ok" }, readiness: { status: "failed", checks: {} } },
      }),
      launcherPid: 42,
      servicePid: () => 84,
    });

    const heartbeat = JSON.parse(await fs.readFile(heartbeatPath, "utf8")) as Record<string, unknown>;
    expect(heartbeat.status).toBe("failed");
    expect(heartbeat.service_running).toBe(false);
  });

  test("launches only a validated external helper for interrupted recovery", async () => {
    const parent = await tempRoot("recover");
    const installRoot = path.join(parent, "RayleaBot");
    const transactionRoot = path.join(parent, ".rayleabot-update-valid");
    await fs.mkdir(installRoot);
    await fs.mkdir(transactionRoot);
    await fs.writeFile(path.join(transactionRoot, "raylea-updater.exe"), "helper");
    await fs.writeFile(path.join(transactionRoot, "journal.json"), JSON.stringify({
      version: 1,
      state: "installing",
      phase: "swap",
      install_root: installRoot,
      transaction_root: transactionRoot,
    }));
    const recoveryLauncher = { spawnDetached: vi.fn(async () => undefined) };

    await expect(launchInterruptedUpdateRecovery(installRoot, 123, recoveryLauncher)).resolves.toBe(true);
    expect(recoveryLauncher.spawnDetached).toHaveBeenCalledWith(
      path.join(transactionRoot, "raylea-updater.exe"),
      ["recover", "--transaction-root", transactionRoot, "--launcher-pid", "123"],
    );
  });
});

File: launcher/internal/desktop/update_resume_test.go
package desktop

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"
)

func TestConsumeUpdateHeartbeatEnvironment(t *testing.T) {
	heartbeatPath := filepath.Join(t.TempDir(), "postflight-launcher-heartbeat.json")
	t.Setenv("RAYLEA_UPDATE_HEARTBEAT", heartbeatPath)
	t.Setenv("RAYLEA_UPDATE_TOKEN", strings.Repeat("g", 64))
	t.Setenv("RAYLEA_UPDATE_RESTART_SERVICE", "true")
	request, present := ConsumeUpdateHeartbeatEnvironment()
	if request != nil || !present {
		t.Fatalf("invalid token request = %#v, present = %v", request, present)
	}
	if os.Getenv("RAYLEA_UPDATE_TOKEN") != "" {
		t.Fatal("update token remained in process environment")
	}

	t.Setenv("RAYLEA_UPDATE_HEARTBEAT", heartbeatPath)
	t.Setenv("RAYLEA_UPDATE_TOKEN", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
	t.Setenv("RAYLEA_UPDATE_RESTART_SERVICE", "true")
	request, present = ConsumeUpdateHeartbeatEnvironment()
	if !present || request == nil || !request.RestartService {
		t.Fatalf("valid request = %#v, present = %v", request, present)
	}
}

func TestCompleteUpdateHeartbeatWritesTrustedBuildIdentity(t *testing.T) {
	artifactID := launcherArtifactID(runtime.GOOS, runtime.GOARCH)
	if artifactID == "" {
		t.Skip("current platform has no formal Launcher artifact")
	}
	parent := t.TempDir()
	installRoot := filepath.Join(parent, "RayleaBot")
	transactionRoot := filepath.Join(parent, ".rayleabot-update-test")
	writeTestFile(t, filepath.Join(installRoot, "build_info.json"), fmt.Sprintf(
		`{"version":"1.2.3","artifact_id":%q,"update_protocol_version":2}`,
		artifactID,
	))
	if err := os.MkdirAll(transactionRoot, 0o700); err != nil {
		t.Fatal(err)
	}
	heartbeatPath := filepath.Join(transactionRoot, "postflight-launcher-heartbeat.json")
	coordinator := NewCoordinator(installRoot, "", 0, nil)

	CompleteUpdateHeartbeat(installRoot, &UpdateHeartbeatRequest{HeartbeatPath: heartbeatPath, Token: "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"}, coordinator)

	payload, err := os.ReadFile(heartbeatPath)
	if err != nil {
		t.Fatalf("read heartbeat: %v", err)
	}
	var heartbeat map[string]any
	if err := json.Unmarshal(payload, &heartbeat); err != nil {
		t.Fatalf("parse heartbeat: %v", err)
	}
	if heartbeat["status"] != "ready" || heartbeat["version"] != "1.2.3" || heartbeat["artifact_id"] != artifactID {
		t.Fatalf("heartbeat = %#v", heartbeat)
	}
}

func TestFindInterruptedUpdateRecoverySelectsNewestValidTransaction(t *testing.T) {
	parent := t.TempDir()
	installRoot := filepath.Join(parent, "RayleaBot")
	if err := os.MkdirAll(installRoot, 0o755); err != nil {
		t.Fatal(err)
	}
	oldRoot := createRecoveryTransaction(t, installRoot, filepath.Join(parent, ".rayleabot-update-old"), "installing")
	newRoot := createRecoveryTransaction(t, installRoot, filepath.Join(parent, ".rayleabot-update-new"), "installing")
	oldTime := time.Now().Add(-time.Minute)
	if err := os.Chtimes(filepath.Join(oldRoot, "journal.json"), oldTime, oldTime); err != nil {
		t.Fatal(err)
	}

	helper, root, ok := findInterruptedUpdateRecovery(installRoot)
	if !ok || !samePath(root, newRoot) || !samePath(helper, filepath.Join(newRoot, "raylea-updater.exe")) {
		t.Fatalf("findInterruptedUpdateRecovery() = %q, %q, %v", helper, root, ok)
	}
}

func TestFindInterruptedUpdateRecoveryIgnoresTerminalJournal(t *testing.T) {
	parent := t.TempDir()
	installRoot := filepath.Join(parent, "RayleaBot")
	if err := os.MkdirAll(installRoot, 0o755); err != nil {
		t.Fatal(err)
	}
	createRecoveryTransaction(t, installRoot, filepath.Join(parent, ".rayleabot-update-done"), "succeeded")
	if _, _, ok := findInterruptedUpdateRecovery(installRoot); ok {
		t.Fatal("terminal update journal was selected for recovery")
	}
}

func createRecoveryTransaction(t *testing.T, installRoot, root, state string) string {
	t.Helper()
	if err := os.MkdirAll(root, 0o700); err != nil {
		t.Fatal(err)
	}
	writeTestFile(t, filepath.Join(root, "raylea-updater.exe"), "helper")
	journal := updateJournal{Version: 1, State: state, InstallRoot: installRoot, TransactionRoot: root}
	payload, _ := json.Marshal(journal)
	writeTestFile(t, filepath.Join(root, "journal.json"), string(payload))
	return root
}
